model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/UserCacheSession.java

```diff
@@ -589,29 +589,37 @@ public int getUsersCount(RealmModel realm, Map<String, String> params, Set<Strin
         return getDelegate().getUsersCount(realm, params, groupIds);
     }
 
+    private UserModel updateCache(RealmModel realm, UserModel delegate) {
+        if (delegate instanceof UserAdapter) {
+            // just return if the entry is an already cached entry of this provider
+            return delegate;
+        }
+        return cacheUser(realm, delegate, cache.getCurrentRevision(delegate.getId()));
+    }
+
     @Override
     public Stream<UserModel> searchForUserStream(RealmModel realm, String search) {
-        return getDelegate().searchForUserStream(realm, search);
+        return getDelegate().searchForUserStream(realm, search).map(u -> updateCache(realm, u));
```

Comment:
I have a little bit of fear of the scenario, when user storage provider returns some way of "lightweight" user from the "search" operations.

For example some user-storage implementation `myUserStorageProvider` where method `searchForUserStream` returns `UserModel` instances with only filled `id` and `username`. As it is an assumption that if admin is interested about particular user `john`, he will then call subsequent call `getUserById`, which is then supposed to return "full" user with all the attributes filled.

If we do the change in this PR, it means that "lightweight" users will be added to the cache, which could then cause issues as subsequent `getUserById` will then also returns this "lightweight" user...

Could this be an issue?

If it could be an issue, the alternative might be to invalidate the users from cache, which are returned by the `search` operations? This will fix the "inconsistency" problem . At the same time, it will have performance impact (the "search" operations are probably not called very often as those are "admin" operations AFAIK, so not sure how much bad is the performance impact...)

Reply:
Yes, that could be an issue. In that case maybe we should just return the cached user instead of the returned user in the store. Not caching anything in searches.

Reply:
Yes, that is also an option. And IMO it is better than current behaviour with "inconsistent" results returned from `search` operations and from operations like `getUserById` . So my vote is to go that path (if @pedroigor and @sguilhen are fine with that).

Reply:
Modified `searchForUserStream` to return the cached entry if present and not cache new entries.

Reply:
@mposolda @sguilhen I'm moving this draft to normal PR as we decided that this is just a bug and no upgrading notes are needed. :smile: